### PR TITLE
feat(donate): PayPal integration

### LINKED ISSUE #37783: Donate Page PayPal Integration
- [ ] PayPal integration for subscription and onetime donations.

### PR DESCRIPTION
**Pull request by ahmaxed**
the integration flow:

1. user clicks on paypal smart button and directed to paypal account for payment approval
2. on approval the subscription id and subscription plan info sent to our api server for verification
3.  our api calls paypal api and verifies the subscription is active and update db.

room for improvement for donation flow:
1. donations forms should be unified and modularized.
2. paypal donation state and stripe donation state should be unified.
3. /donate route for the api should be split into multiple files 
4. lato should be used for all donation related text for ease of use.
5. donation tests should be added for the api and client
6. donation responses should change client redux store

Also I intentionally avoided adding tests since the client and api of donation will noticeably change in the next week. then we could optimize and add tests if that is ok with everyone.

In order to review this pr you would need the following keys in the .env
PAYPAL_CLIENT_ID
PAYPAL_SECRET
PAYPAL_FIVE_DOLLAR_MONTHLY_PLAN_ID

Closes #37783

### CODE REVIEW

File: api-server/common/models/user.js [hunk 1]
```diff
@@ -439,6 +439,10 @@ export default function(User) {
     )(donation).do(() =>
       this.updateAttributes({
         isDonating: true,
+        donationProviders: [
```
Comment (raisedadead) [inline api-server/common/models/user.js]: What happens when you have three stripe donations, and two PayPal donations? Do we want to track repetitions? If so why?

File: api-server/server/boot/donate-paypal.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
+/* eslint-disable max-len */
```
Comment (raisedadead) [inline api-server/server/boot/donate-paypal.js]: Lets fix the cause for this, and not the symptom.

Comment (raisedadead) [reply api-server/server/boot/donate-paypal.js]: Also, why do we need additional boot script for PayPal?

File: api-server/server/models/donation.json [hunk 3]
```diff
@@ -45,7 +45,6 @@
     },
     "customerId": {
       "type": "string",
-      "required": true,
```
Comment (raisedadead) [inline api-server/server/models/donation.json]: Is this change intentional?

File: config/env.js [hunk 4]
```diff
@@ -20,7 +20,9 @@ const {
   STRIPE_PUBLIC_KEY: stripePublicKey,
   SERVICEBOT_ID: servicebotId,
   ALGOLIA_APP_ID: algoliaAppId,
-  ALGOLIA_API_KEY: algoliaAPIKey
+  ALGOLIA_API_KEY: algoliaAPIKey,
+  PAYPAL_CLIENT_ID: paypalClientId,
+  PAYPAL_FIVE_DOLLAR_MONTHLY_PLAN_ID: paypalFiveDollarMonthlyPlanId
```
Comment (raisedadead) [inline config/env.js]: Whats the plan when we add 15 different plans?

File: sample.env [hunk 5]
```diff
@@ -23,6 +23,9 @@ SERVICEBOT_ID=servicebot_id_from_servicebot_dashboard
 SERVICEBOT_HMAC_SECRET_KEY=secret_key_from_servicebot_dashboard
 
 PAYPAL_SUPPORTERS=1703
+PAYPAL_CLIENT_ID=id_from_paypal_dashboard
+PAYPAL_SECRET=secret_from_paypal_dashboard
+PAYPAL_FIVE_DOLLAR_MONTHLY_PLAN_ID=subscription_plan_id_from_paypal_dashboard
```
Comment (raisedadead) [inline sample.env]: does the subscription id require us to keep it secret? If not, whats the best place to keep it?

**Review by raisedadead — COMMENTED** [review]
I haven't gone through the client code yet. You say this (whole implementation, I assume) is going to change in upcoming days, can you elaborate?

**Comment by raisedadead** [comment]
> on approval the subscription id and subscription plan info sent to our api server for verification
> our api calls paypal api and verifies the subscription is active and update db.

I realised that we are unable to proceed as is. We would be needing the web-hooks, because with this we are unable to track cancelations done directly with PayPal.

As a user, I am able to login to PayPal and directly cancel my donation. How are we handling that use case.

This is something we lack in Stripe as well and we should fix this for both.

File: api-server/common/models/user.js (hunk 1 above)
Comment (ahmadabdolsaheb) [reply api-server/common/models/user.js]: I am not entirely sure that we need donationProviders in the user object to be honest since the users presumably are prevented from making multiple donations, and we can pull the donation related info from the donation collection.

File: api-server/server/models/donation.json (hunk 3 above)
Comment (ahmadabdolsaheb) [reply api-server/server/models/donation.json]: This, since the customer id is not provided when we get the subscription info from paypal.

File: api-server/server/boot/donate-paypal.js (hunk 2 above)
Comment (ahmadabdolsaheb) [reply api-server/server/boot/donate-paypal.js]: just to reduce the size of the donate I moved the paypal related logic to another file which could probably be placed in a helper directory.

Please let me know of your preferred way to proceed.

File: config/env.js (hunk 4 above)
Comment (ahmadabdolsaheb) [reply config/env.js]: So the plan id changes between seeding and live that is why we cannot hard code it. since we have a mechanism to distribute the keys and secrets from env I thought it might be a good idea to keep it here.

Please let me know if a better idea comes to mind.

File: sample.env (hunk 5 above)
Comment (ahmadabdolsaheb) [reply sample.env]: the subscription id changes between sandbox and live. so I thought it would better to keep it in env where we can change it easily for dev and production

**Comment by ahmadabdolsaheb** [comment]
> > on approval the subscription id and subscription plan info sent to our api server for verification
> > our api calls paypal api and verifies the subscription is active and update db.
> 
> I realised that we are unable to proceed as is. We would be needing the web-hooks, because with this we are unable to track cancelations done directly with PayPal.
> 
> As a user, I am able to login to PayPal and directly cancel my donation. How are we handling that use case.
> 
> This is something we lack in Stripe as well and we should fix this for both.

I can add that for cancellation to this PR.

**Comment by ahmadabdolsaheb** [comment]
> I haven't gone through the client code yet. You say this (whole implementation, I assume) is going to change in upcoming days, can you elaborate?


at this point the minimalDonationForm has a state that is in charge of paypal submission.
On the other hand DonateFormChildViewForHOC which is nested in MinimalDonationForm and DonateForm has a state that is in charge of the stripe from submission.

Ideally we would need to combine MinimalDonationForm and DonateForm into one file and make  it in charge of all donation state ( stripe, paypal, apple pay, amazon pay, ....), then all nested components such as DonateFormChildViewForHOC would report their state upwards. That way we would have a single source of truth for all payment submissions. 

That being said, the current setup works but it is not optimal. Therefore I would prefer having tests after we land the above-mentioned improvement.

File: api-server/server/models/donation.json (hunk 3 above)
Comment (raisedadead) [reply api-server/server/models/donation.json]: What would be the alternate from PayPal?

File: config/env.js (hunk 4 above)
Comment (raisedadead) [reply config/env.js]: I understand it is an identifier and hence it changes. Is it required to be kept a secret? Is it a configuration entry? Can we store this somewhere else? Have we done this already with Stripe's amounts?

File: sample.env (hunk 5 above)
Comment (raisedadead) [reply sample.env]: See https://github.com/freeCodeCamp/freeCodeCamp/pull/38239#discussion_r384335418

**Comment by raisedadead** [comment]
> I can add that for paypal cancellation to this PR.

You mean you can add web-hooks?

File: api-server/server/models/donation.json (hunk 3 above)
Comment (ahmadabdolsaheb) [reply api-server/server/models/donation.json]: for the subscriber the only provide name, email and address.

**Comment by ahmadabdolsaheb** [comment]
> > I can add that for paypal cancellation to this PR.
> 
> You mean you can add web-hooks?

yes.

**Comment by raisedadead** [comment]
> at this point the ...

That's all client side? So the API side changes are not changing. If yes, let's get some tests in.

**Comment by raisedadead** [comment]
> yes.

If we are going to be adding web-hooks for cancellations, then we should be doing it for sending the accepting the donations as well. The point being if you are to be implementing calls to the PayPal's API endpoints why not make it single source too. 

We wanted the validation to begin with.

File: api-server/common/models/user.js (hunk 1 above)
Comment (raisedadead) [reply api-server/common/models/user.js]: Okay, in that case, you can go ahead and remove all this logic.

File: api-server/server/boot/donate-paypal.js (hunk 2 above)
Comment (raisedadead) [reply api-server/server/boot/donate-paypal.js]: Loopback hooks the boot scripts in the boot folder (presumably in alphabetical order). The size of the file does not matter on the server side of things unless it is meant to be shared across other places.

File: api-server/server/models/donation.json (hunk 3 above)
Comment (raisedadead) [reply api-server/server/models/donation.json]: So, what would be the alternate uniquely identifiable information / item that we can query from their API while making calls to them. For instance, if we wanted to get all the transactions for a person. If it's the email, could hat be used instead?

**Comment by ahmadabdolsaheb** [comment]
> > at this point the ...
> 
> That's all client side? So the API side changes are not changing. If yes, let's get some tests in.

Sure. By tests you are referring to api tests or client and api?

File: config/env.js (hunk 4 above)
Comment (ahmadabdolsaheb) [reply config/env.js]: the sub id is not a secret and is only used when paypal button is clicked to create subscriptions.

I could hard code both ids in config.js and add the logic to write the correct one to config.json during build. how does that sound?

Comment (raisedadead) [reply config/env.js]: Which `config.js`?

**Comment by ahmadabdolsaheb** [comment]
> If we are going to be adding web-hooks for cancellations, then we should be doing it for accepting the donations as well. The point being if you are to be implementing calls to the PayPal's API endpoints why not make it single source too.
> 
> We wanted the validation to begin with.

I am not sure what exactly you are referring to by validation. do you mean making sure the subscription exist? 

If so based on my experience the web-hooks are not fired in a timely manner meaning we are not able to update the UI appropriately if we wait for hooks and then validate a subscription.

Ideally, we need to selectively capture the subscription payments or suspend them (ex. payment already captured on paypal side) if then subscription does not meet our requirements. Additionally  we could utilize web-hooks for events that we do not have control over such as suspension or cancelation by the user.

Let me know if I am way off base here :)

File: config/env.js (hunk 4 above)
Comment (ahmadabdolsaheb) [reply config/env.js]: sorry env.js in /config  https://github.com/freeCodeCamp/freeCodeCamp/blob/master/config/env.js

**Comment by ahmadabdolsaheb** [comment]
for now, I will do the following improvements:

1. send the subscription id to the backend after user approval and then active in the backend. (if our tests pass)
2. add hooks for cancelation( would you like to add a status to donation records and set them to "cancelled"?)
3. add test to the backend ( after finalizing the logic ).

I can also implement the state unification on the client and add tests for that as well if you wish?

File: config/env.js (hunk 4 above)
Comment (raisedadead) [reply config/env.js]: > the sub id is not a secret and is only used when paypal button is clicked to create subscriptions.

Since, its not a secret why can't it be static instead. Did you get a chance to look at: 

https://github.com/freeCodeCamp/freeCodeCamp/blob/master/config/donation-settings.js

**Comment by raisedadead** [comment]
> I am not sure what exactly you are referring to by validation.

All the interaction with PayPal's API ideal should go thru our API, like we already do for Stripe. That gives us a chance to implement limits, flexibility, control and more.

> If so based on my experience the web-hooks are not fired in a timely manner meaning we are not able to update the UI appropriately if we wait for hooks and then validate a subscription.
> 
> Ideally, we need to selectively capture the subscription payments or suspend them (ex. payment already captured on paypal side) if then subscription does not meet our requirements. Additionally we could utilize web-hooks for events that we do not have control over such as suspension or cancelation by the user.
> 
> Let me know if I am way off base here :)

Agreed. Web-hooks are async. 

This also means you would implement logic to handle cancellations, etc. meaning adding APIs to interact with PayPal's APIs.

Why would we not add APIs to create the subscription in the first place.

For all I know the "Smart Buttons" are creating a id that we are sending from our client app to our API, to verify and update the db on our side.

We do not need that cruft, PayPal would be delivering the web-hook when the subscription was made. 

You can safely assume that they are delivered with the same safety with the async logic that you would be using to verify the transaction from the smart button.

What I think is that we should use that onApprove hook to simply show the user that the donation was approved and we would have the status of the same shortly. It is a quick enough message to avoid making the UX blocking.

**Comment by ahmadabdolsaheb** [comment]
> All the interaction with PayPal's API ideal should go thru our API, like we already do for Stripe. That gives us a chance to implement limits, flexibility, control and more.

To my understanding, since we do not want the users to be redirected to another page we are using smart buttons that come with their own limitations (features). We could make the smart button create subscriptions with an 'approved' status which we could validate and then activate from our own api. 

Although now that think about it we should not reject any subscription with paypal because paypal will bear the burden of all fraud. so even if the user is not signed in or has multiple donation we should still allow them to subscribe.

> We do not need that cruft, PayPal would be delivering the web-hook when the subscription was made.

Updating the db with hooks would simplify things. Although we would still need a small workaround since the webhooks does not include any user info with which we could update the related user records. we would still need to send some data from smart buttons to our api and create some soft of records with the subscriber email so when the webhook arrives we could look up the user id from that record and update the user accordingly. 

In that case, it would be better to directly update the user records when the data arrives from client.

File: config/env.js (hunk 4 above)
Comment (ahmadabdolsaheb) [reply config/env.js]: yes, I did. they could be static but they will change based on environment. so somewhere I will need to check for the production environment and set the right id to be used by the client.

File: api-server/server/boot/donate-paypal.js (hunk 2 above)
Comment (ahmadabdolsaheb) [reply api-server/server/boot/donate-paypal.js]: I will include all donation related logic in the same file then.

File: api-server/server/models/donation.json (hunk 3 above)
Comment (ahmadabdolsaheb) [reply api-server/server/models/donation.json]: we can get transactions of a subscription id but not a specific person.
https://developer.paypal.com/docs/api/subscriptions/v1/#subscriptions_transactions

**Comment by raisedadead** [comment]
> To my understanding, since we do not want the users to be redirected to another page we are using smart buttons that come with their own limitations (features).

There is no re-direction involved. The onApproved callback is the same callback that you would have used to send the id back to our API. Instead, you would use the same to show message? Right?

> it would be better to directly update the user records when the data arrives from client.

That is the whole point. Since that data is coming from the client. It is not trusted. It can be spoofed. Correct?

If you are going to be using the id from the button to fetch and update the data from PayPal, you are able to do the same from the hook as well with whatever identifier thy gave. It is literally the same implementation.

The smart buttons are a solution made for merchants who want a quick way of integrating product and subscription sales. It does not come with the tracking features that we want.

Am I missing something here?

**Comment by raisedadead** [comment]
Here is what the workflow should be:

- First the user lands on the page where they see the PayPal button (smart or our own, I will let you figure that out).
- They click through the button and are taken thru the PayPal flow of capturing a transaction.
- The client SDK or our own button have the callback from the previous step.
- We show a success message.
- The API gets pinged with hook, and we update the records as needed (subscription/cancellation) etc.

**Comment by ahmadabdolsaheb** [comment]
> * The API gets pinged with hook, and we update the records as needed (subscription/cancellation) etc.

I am not clear about the last step. the hook data only includes subscription id, plan id, subscriber name and address, email, etc. it does not include any information with which we could find a user and update the db accordingly. that is my confusion.

**Comment by raisedadead** [comment]
And that is precisely the reason I was looking for a way to avoid the smart buttons and drive the payment thru our API for control.

There are some compromises to be made here, which is (definitely not security), we should tally by create/update a user record with the email that we receive from PayPal.

If a user choses to use a different email than on record on PayPal for whatever reason, and the email is not with us, we create a new user record. (This is same as the current Stripe flow).

**Comment by ahmadabdolsaheb** [comment]
> If a user choses to use a different email than on record on PayPal for whatever reason, and the email is not with us, we create a new user record. (This is same as the current Stripe flow).

Forgive for my confusion, it seems like we are modifying the signed in user that is fed to createStripeDonation function here:
https://github.com/freeCodeCamp/freeCodeCamp/blob/059e826465b789d4c76bbd13525fd6b9a1859f03/api-server/server/boot/donate.js#L102

I do not see where we create new users upon email mismatch. 
I will investigate more and if you could direct me to the right block, it is highly appreciated.

**Comment by raisedadead** [comment]
Are you going to be using smart buttons or the API?

Like I said,

> And that is precisely the reason I was looking for a way to avoid the smart buttons and drive the payment thru our API for control.
> 
> Anyway, since we are not going that way, there are some compromises to be made here, which is (definitely not security), we should tally by create/update a user record with the email that we receive from PayPal.
> 
> If a user choses to use a different email than on record on PayPal for whatever reason, and the email is not with us, we create a new user record. (This is same as the current Stripe flow).

I thought you are using the Smart Buttons from PayPal. If Yes, this means you have two things to do:

1. Use the onApprove callback to show the message on client and nothing else.
2. Followed by use the web-hook delivered to our API (which has nothing to do with sign in state, because its not happening on the client) to create or update a user record. If a user used a different email id on PayPal than with us, it's going to mean a new record.

Does that help? Let me know.

**Comment by ahmadabdolsaheb** [comment]
It makes sense now.

**Comment by ojeytonwilliams** [comment]
So, my understanding is that PayPal is responsible for preventing fraud, so all we need to do is track donations on the server.  Given that, shouldn't the 'create a donation' flow look like this:

- Donate button is pressed
- PayPal window opens, user creates a donation
- PayPal tells the client (not the server) when this is done
- The client displays a message "Payment confirmed, updating our records" or similar
- The client communicates the donation information to the server, with an attached `user` object
- The server checks to see if the plan is valid i.e. is PAYPAL_FIVE_DOLLAR_MONTHLY_PLAN_ID
- The server makes sure the id is not currently in the database
- The server calls the PayPal api to verify that the donation exists
    - If the donation exists then
        - If that `user` exists on the server update it to track the donation    
        - If not, create *something* to track the donation, warn the devs and perhaps the user
        
    - If the donation does not exist then there's (probably) no donation to track, but the information should be logged somewhere in case of problems and the user informed that something went wrong
- Finally the server reports back to the client and the user gets a confirmation message
     

The server call to PayPal would be instead of a webhook, since it verifies the id is valid and the plan is one of ours.  I think that's all it needs to do.

**Comment by raisedadead** [comment]
And you would be correct, but like I said this earlier in the thread, but the thing is we would still depend on the web-hooks for cancellation scenarios happening on PayPal.

This means we would be writing code to handle the hooks, and if we are doing that already, we can do it for subscription creation. It simpler to maintain.

There is no client side to server to client roundtrips required.

**Comment by ojeytonwilliams** [comment]
Understood.  My main worry is the case where the user's PayPal email address is different from their fcc address and we can't find them.  Creating a separate user to track the donation will probably cause some issues.  For instance the existing user should get the `isDonating` flag, but they wouldn't.

Possible solutions: 
  - add a flag to donation user accounts so, if there are problems, we at least know which ones are affected
  - trust the client enough to allow it to set the `isDonating` flag on successful donation.

I'm not arguing against webhooks, mind, I appreciate the need.  I'm just trying to think through the consequences.

**Comment by raisedadead** [comment]
> I'm just trying to think through the consequences.

Yeah, I could not agree with you more.

We could update the `isDonating` flag regardless, because the onApprove callback means the donation was captured.

**Comment by ojeytonwilliams** [comment]
> We could update the `isDonating` flag regardless, because the onApprove callback means the donation was captured.

Oh, of course, that's much nicer.

**Comment by ahmadabdolsaheb** [comment]
@raisedadead, could you please eyeball this solution so I could go ahead with the tests? thank you.

File: api-server/server/boot/donate.js [hunk 6]
```diff
@@ -246,24 +258,89 @@ export default function donateBoot(app, done) {
...
   const publicInvalid = !pubKey || pubKey === 'pk_from_stripe_dashboard';
+
+  const paypalSecretInvalid =
+    !paypalKey || paypalKey === 'id_from_paypal_dashboard';
+  const paypalPublicInvalid =
+    !paypalSec || paypalSec === 'secret_from_paypal_dashboard';
   const hmacKeyInvalid =
     !hmacKey || hmacKey === 'secret_key_from_servicebot_dashboard';
+  const paypalInvalid = paypalPublicInvalid || paypalSecretInvalid;
+  const stripeInvalid = secretInvalid || publicInvalid || hmacKeyInvalid;
```
Comment (raisedadead) [inline api-server/server/boot/donate.js]: These should be explicitly renamed to refer to stripe. Also since the HMAC key comes from servicebot, a separate service we should be treating it as such.

File: api-server/server/boot/donate.js [hunk 7]
```diff
@@ -246,24 +258,89 @@ export default function donateBoot(app, done) {
...
 
-  if (secretInvalid || publicInvalid || hmacKeyInvalid) {
+  if (stripeInvalid) {
     if (process.env.FREECODECAMP_NODE_ENV === 'production') {
       throw new Error('Stripe API keys are required to boot the server!');
     }
     console.info('No Stripe API keys were found, moving on...');
-    done();
```
Comment (raisedadead) [inline api-server/server/boot/donate.js]: `done` is being removed, where is this handled?

File: api-server/server/middlewares/csurf.js [hunk 8]
```diff
@@ -7,8 +7,11 @@ export default function() {
     }
   });
   return function csrf(req, res, next) {
-    const path = req.path.split('/')[1];
-    if (/(^api$|^unauthenticated$|^internal$|^p$)/.test(path)) {
+    const pathArr = req.path.split('/');
+    if (/(^api$|^unauthenticated$|^internal$|^p$)/.test(pathArr[1])) {
+      return next();
+    }
+    if (/(^donate$)/.test(pathArr[1]) && /(^update-paypal$)/.test(pathArr[2])) {
```
Comment (raisedadead) [inline api-server/server/middlewares/csurf.js]: Why is this change required?

File: api-server/server/models/donation.json (hunk 3 above)
Comment (raisedadead) [reply api-server/server/models/donation.json]: I would re-visit this, and add an equivalent.

File: client/src/components/Donation/DonationModal.js [hunk 9]
```diff
@@ -88,8 +92,8 @@ function DonateModal({
 
   const donationText = (
     <b>
-      Become a supporter and help us create even more learning resources for
-      you.
+      Become a $5 / month supporter and help us create even more learning
```
Comment (raisedadead) [inline client/src/components/Donation/DonationModal.js]: I do not like the idea of static amounts in code. Quincy always gets us to keep tweaking this, please pull this from the config.

File: config/secrets.js [hunk 10]
```diff
@@ -33,7 +33,10 @@ const {
   STRIPE_PUBLIC_KEY,
   STRIPE_SECRET_KEY,
   SERVICEBOT_ID,
-  SERVICEBOT_HMAC_SECRET_KEY
+  SERVICEBOT_HMAC_SECRET_KEY,
```
Comment (raisedadead) [inline config/secrets.js]:
```suggestion

  SERVICEBOT_HMAC_SECRET_KEY,
```

Comment (raisedadead) [reply config/secrets.js]: newline

**Review by raisedadead — CHANGES_REQUESTED** [review]
Quick Comments.

File: api-server/server/boot/donate.js (hunk 7 above)
Comment (ojeytonwilliams) [reply api-server/server/boot/donate.js]: It's handled down here:
https://github.com/freeCodeCamp/freeCodeCamp/blob/d62e37b2aaaa421ea1855d0a02bfdc8a3bc7418a/api-server/server/boot/donate.js#L341-L344

I think we should move `addDonation` and `updatePaypal` out of the way, since defining them in the middle of `donateBoot` makes it harder to follow the logic.

File: api-server/server/middlewares/csurf.js (hunk 8 above)
Comment (ahmadabdolsaheb) [reply api-server/server/middlewares/csurf.js]: Although csurf helps protect against cross-site request forgery attempts, I believe it also prevents us from processing legitimate webhook events. That is why I exempted the webhook route from CSRF protection.

That being said, we verify each events that hits that route and make sure they come from paypal before proceeding with modifying our records.

File: client/src/components/Donation/DonationModal.js (hunk 9 above)
Comment (ahmadabdolsaheb) [reply client/src/components/Donation/DonationModal.js]: I will tweak the config and the UI in an upcoming pr. if that is alright with you, it would be better to handle it there.

File: api-server/server/models/donation.json (hunk 3 above)
Comment (ahmadabdolsaheb) [reply api-server/server/models/donation.json]: I added an activation mock hook with its headers and body. There does not seem to be an equivalent. 
I could re add the email as customerId to satisfy this requirement.

Comment (raisedadead) [reply api-server/server/models/donation.json]: Yup, adding the email should be fine.

File: client/src/components/Donation/PaypalButton.js [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    };
+    this.handleApproval = this.handleApproval.bind(this);
+  }
+
+  handleApproval = data => {
+    this.props.handleProcessing('month', 500, 'Paypal payment submission');
+    this.props.onDonationStateChange(false, true, '');
+    verifySubscriptionPaypal(data)
+      .then(response => {
+        console.log(response.data);
```
Comment (ojeytonwilliams) [inline client/src/components/Donation/PaypalButton.js]:
```suggestion
```

File: api-server/server/boot/donate.js [hunk 12]
```diff
@@ -243,27 +255,95 @@ export default function donateBoot(app, done) {
...
-    app.use(donateRouter);
-    app.use('/internal', donateRouter);
+  }
+  if (paypalInvalid) {
+    if (process.env.FREECODECAMP_NODE_ENV === 'production') {
+      throw new Error('PayPal API keys are required to boot the server!');
+    }
+    console.info('No PayPal API keys were found, moving on...');
+  } else {
+    api.post('/update-paypal', updatePaypal);
```
Comment (ojeytonwilliams) [inline api-server/server/boot/donate.js]: This endpoint is not being called yet and so the user record does not get updated.  I think it would make sense if this was called by the client and then, if the user update succeeds, a new donation record should be created (via the logic that's currently being triggered by `add-donation`)

File: utils/donation.js [hunk 13]
```diff
@@ -0,0 +1,7 @@
+exports.getDonation = function getDonation(name) {
+  return ('' + name)
+    .toLowerCase()
+    .trim()
+    .replace(/\s/g, '-')
+    .replace(/[^a-z\d\-.]/g, '');
+};
```
Comment (ojeytonwilliams) [inline utils/donation.js]: Is this used?

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
It is currently working, insofar as the donations are getting created as expected.  However, the verification endpoint is not getting used and the user's donation emails are not getting updated.

File: api-server/server/boot/donate.js (hunk 12 above)
Comment (ahmadabdolsaheb) [reply api-server/server/boot/donate.js]: so you should tell paypal to send the webhooks to a specific endpoint. I have it setup to my device at the moment but we will be able to point it to .dev and .org for testing.

Comment (ojeytonwilliams) [reply api-server/server/boot/donate.js]: Good point. I completely misunderstood this again.  Sorry about that!

So, for QAing, can we set this up on .dev with this PR and with `FREECODECAMP_NODE_ENV`  set to `development`?

File: utils/donation.js (hunk 13 above)
Comment (ahmadabdolsaheb) [reply utils/donation.js]: no, not sure why it is there to tbh :)

File: api-server/server/boot/donate.js (hunk 12 above)
Comment (ahmadabdolsaheb) [reply api-server/server/boot/donate.js]: Yes we just need to update the donation settings in the config and change the keys in the env

**Comment by ahmadabdolsaheb** [comment]
The only consideration is when /update-paypal receives calls that are not hooks, it first returns 200 and then errors out. looking into it ATM

File: config/donation-settings.js [hunk 14]
```diff
@@ -39,6 +39,30 @@ const donationSubscriptionConfig = {
...
+  production: {
+    webhookId: '',
+    durationPlans: {
+      month: {
+        '500': {
+          planId: ''
+        }
+      }
+    }
+  },
```
[the remark this answers is not in the document]
Comment (ahmadabdolsaheb) [reply config/donation-settings.js]: Yes, since they are going to be long list of plan ids, we decided to keep them all in the donation-settings and display the right one based on the env variable and the donation amount.

[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply config/donation-settings.js]: If I understand this correctly, `paypalConfig.production` is what gets used, but that doesn't (currently) have a `webhookId`.  So, is that property going to be overwritten by environment variables?

[the remark this answers is not in the document]
Comment (ahmadabdolsaheb) [reply config/donation-settings.js]: no I just don't have that yet :). I will just add it when we get everything setup at the end.

File: api-server/server/utils/donation.js [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+    .createDonation(donation)
+    .toPromise()
+    .catch(err => {
+      throw new Error(err);
+    });
+};
+
+export function removeFirstEmail(arr, email) {
+  let index = arr.indexOf(email);
+  if (index < 0) return arr;
```
[the remark this answers is not in the document]
Comment (ahmadabdolsaheb) [reply api-server/server/utils/donation.js]: Added a log for that. thanks for pointing it out

File: api-server/server/middlewares/csurf.js (hunk 8 above)
Comment (raisedadead) [reply api-server/server/middlewares/csurf.js]: We can unify the same into the previous check :)

Comment (raisedadead) [reply api-server/server/middlewares/csurf.js]: As discussed we do not need this anymore, but please add a comment block stating that any changes to this must accommodate for the fact that PayPal hits these without CSRF.

File: api-server/server/utils/donation.js [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+        message: `Failed token verification.`,
+        type: 'FailedPaypalTokenVerificationError'
+      };
+    }
+  } catch (err) {
+    if (err.type === 'FailedPaypalTokenVerificationError') {
+      throw err;
+    }
+    throw new Error(err);
+  }
```
[the remark this answers is not in the document]
Comment (ahmadabdolsaheb) [reply api-server/server/utils/donation.js]: try catches has been removed as suggested. thanks

File: client/src/components/Donation/MinimalDonateForm.js [hunk 17]
```diff
@@ -91,7 +148,7 @@ class MinimalDonateForm extends Component {
                 donationAmount={donationAmount}
                 donationDuration={donationDuration}
                 getDonationButtonLabel={() =>
-                  `Confirm your donation of $5 per month`
+                  `Confirm your donation of $5 / month`
```
Comment (raisedadead) [inline client/src/components/Donation/MinimalDonateForm.js]: Lets pick this from the config or something.

File: api-server/server/boot/donate.js [hunk 18]
```diff
@@ -243,24 +255,73 @@ export default function donateBoot(app, done) {
...
+  const stripeInvalid = stripeSecretInvalid || stripPublicInvalid;
 
-  if (secretInvalid || publicInvalid || hmacKeyInvalid) {
+  if (stripeInvalid || paypalInvalid || hmacKeyInvalid) {
     if (process.env.FREECODECAMP_NODE_ENV === 'production') {
-      throw new Error('Stripe API keys are required to boot the server!');
+      throw new Error('Donation API keys are required to boot the server!');
     }
-    console.info('No Stripe API keys were found, moving on...');
+    log('Some Donation API keys were not found, moving on...');
```
Comment (ojeytonwilliams) [inline api-server/server/boot/donate.js]:
```suggestion
    log('Donation disabled in development unless ALL test keys are provided');
```
Just to remind people they can't test, say, PayPal with only those keys.

File: api-server/server/boot/donate.js [hunk 19]
```diff
@@ -244,23 +250,73 @@ export default function donateBoot(app, done) {
...
+    const { headers, body } = req;
+    return Promise.resolve(req)
+      .then(verifyWebHookType)
+      .then(getAsyncPaypalToken)
+      .then(token => verifyWebHook(headers, body, token, keys.paypal.webhookId))
+      .then(hookBody => updateUser(hookBody, app))
+      .then(() => res.status(200).json({ message: 'received hook' }))
+      .catch(err => {
+        log(err.message);
+        return res.status(200).json({ message: 'received hook' });
```
Comment (ojeytonwilliams) [inline api-server/server/boot/donate.js]: Is the plan to manually check the logs to see what the problem was, rather than to get PayPal to try again?  If so, I think the messages will need more information to allow you to find which donation went wrong.

Also, the reply can go in `finally` if it's always the same.

Comment (raisedadead) [reply api-server/server/boot/donate.js]: Well I think we should be hooking up something similar to rollbar for this. And the reply is lost by PayPal, I did not say a way that they log it. So we should depend on our logging for this.

**Review by raisedadead — APPROVED** [review]

File: api-server/server/boot/donate.js (hunk 19 above)
Comment (raisedadead) [reply api-server/server/boot/donate.js]: I created an issue to track your feedback.
